tensorstore/kvstore/s3/aws_metadata_credential_provider.cc

```diff
@@ -0,0 +1,179 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aws_metadata_credential_provider.h"
+
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/internal/json_binding/bindable.h"
+#include "tensorstore/internal/json_binding/json_binding.h"
+
+using ::tensorstore::Result;
+using ::tensorstore::internal::ParseJson;
+using ::tensorstore::internal_http::HttpResponseCodeToStatus;
+using ::tensorstore::internal_kvstore_s3::AwsCredentials;
+
+namespace jb = tensorstore::internal_json_binding;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// EC2 Metadata server url
+static constexpr char kEc2MetadataUrl[] = "http://http://169.254.169.254";
+// Token ttl header
+static constexpr char kTokenTtlHeader[] = "x-aws-ec2-metadata-token-ttl-seconds";
+// Token header
+static constexpr char kMetadataTokenHeader[] = "x-aws-ec2-metadata-token";
+
+static constexpr char kTokenPath[] = "/latest/api/token";
+static constexpr char kIamPath[] = "/latest/meta-data/iam/";
+static constexpr char kIamSecurityCredentialsPath[] = "/latest/meta-data/iam/security-credentials/";
+
+// Requests to the above server block outside AWS
+// Configure a timeout small enough not to degrade performance outside AWS
+// but large enough inside AWS
+static constexpr absl::Duration kConnectTimeout = absl::Milliseconds(200);
+
+/// Represents JSON returned from
+/// http://http://169.254.169.254/latest/meta-data/iam/security-credentials/<iam-role>/
+/// where <iam-role> is usually returned as a response from a request to
+/// http://http://169.254.169.254/latest/meta-data/iam/security-credentials/
+struct EC2CredentialsResponse {
+  std::string Code;
+  absl::Time LastUpdated;
+  std::string Type;
+  std::string AccessKeyId;
+  std::string SecretAccessKey;
+  std::string Token;
+  absl::Time Expiration;
+
+  using ToJsonOptions = IncludeDefaults;
+  using FromJsonOptions = internal_json_binding::NoOptions;
+
+  TENSORSTORE_DECLARE_JSON_DEFAULT_BINDER(
+      EC2CredentialsResponse,
+      internal_kvstore_s3::EC2CredentialsResponse::FromJsonOptions,
+      internal_kvstore_s3::EC2CredentialsResponse::ToJsonOptions)
+};
+
+inline constexpr auto EC2CredentialsResponseBinder = jb::Object(
+    jb::Member("Code", jb::Projection(&EC2CredentialsResponse::Code)),
+    jb::Member("LastUpdated", jb::Projection(&EC2CredentialsResponse::LastUpdated)),
+    jb::Member("Type", jb::Projection(&EC2CredentialsResponse::Type)),
+    jb::Member("AccessKeyId", jb::Projection(&EC2CredentialsResponse::AccessKeyId)),
+    jb::Member("SecretAccessKey", jb::Projection(&EC2CredentialsResponse::SecretAccessKey)),
+    jb::Member("Token", jb::Projection(&EC2CredentialsResponse::Token)),
+    jb::Member("Expiration", jb::Projection(&EC2CredentialsResponse::Expiration))
```

Comment:
I suspect that, except for `Code`, all of these should be optional to encourage success in json parsing, when `Code != "Success"`.

A sample  from https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials

```json
{
  "Code" : "Success",
  "LastUpdated" : "2012-04-26T16:39:16Z",
  "Type" : "AWS-HMAC",
  "AccessKeyId" : "ASIAIOSFODNN7EXAMPLE",
  "SecretAccessKey" : "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
  "Token" : "token",
  "Expiration" : "2017-05-17T15:09:54Z"
}
```

I'll try find the schema.

Reply:
I've made all members optional except  except `Code` and added a test case for the `Code == SomethingOtherThanSuccess` case.